src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp

```diff
@@ -320,17 +321,48 @@ void GenericOTARequestorDriver::StopDefaultProviderTimer()
         this);
 }
 
-// Returns the next available Provider location
+/**
+ * Returns the next available Provider location. The algorithm is to simply loop through the list of DefaultOtaProviders and return
+ * the next value (based on the last used provider). If no suitable candidate is found, FALSE is returned.
+ */
+
 bool GenericOTARequestorDriver::DetermineProviderLocation(ProviderLocation::Type & providerLocation)
 {
+    // Iterate through the default providers list and find the last used provider. If found, return the provider after it
     auto iterator = mRequestor->GetDefaultOTAProviderListIterator();
-    while (iterator.Next())
+    while (mLastUsedProvider.HasValue() && iterator.Next())
+    {
+        if (ProviderLocationsEqual(iterator.GetValue(), mLastUsedProvider.Value()))
+        {
+            if (iterator.Next())
+            {
+                providerLocation = iterator.GetValue();
+                return true;
+            }
+        }
+    }
+
+    // If no suitable candidate found, return in the following order:
+    // - first element of the default providers list OR
+    // - the last used provider OR
+    // - error (no suitable candidate)
+    iterator = mRequestor->GetDefaultOTAProviderListIterator();
+    if (iterator.Next())
     {
-        // For now, just return the first one
         providerLocation = iterator.GetValue();
         return true;
     }
-
+    else if (mLastUsedProvider.HasValue())
```

Comment:
Why is this needed? It can only be true if the last used provider is no longer among the default OTA providers. I don't think we should keep using it, right?

Reply:
So, this should only happen in the case where the Default OTA Providers List is empty for some reason. This should not happen unless something bad has happened (like an entity wiping the OTA providers list or a bug). In this case, I thought returning the only "valid" OTA provider was better than nothing at all. For all we know this provider might have an update.

Reply:
Imho, if the list is empty, then this function should return false. We shouldn't be using some stale entry.

Reply:
I see your point about the stale entry. Will remove.

Reply:
Last used provider could be the one we received in the AnnounceOTAProvider command, it might not be present in the Default OTA Providers List and we don't add it to that list -- so this is a valid situation. 

Whether we want to keep retrying the provider from AnnounceOTAProvider is a separate question -- in my opinion either choice is fine, may as well retry.

Reply:
If we want to use the provider announced with `AnnounceOTAProvider` then I think we need to additionally mark the last used provider source, so that we don't use a removed default provider.